thoth-api/src/model/mod.rs

```diff
@@ -1,5 +1,152 @@
-#[cfg(feature = "backend")]
-use crate::errors::ThothResult;
+use chrono::{DateTime, TimeZone, Utc};
+use serde::{Deserialize, Serialize};
+use std::fmt;
+use std::str::FromStr;
+
+use crate::errors::{ThothError, ThothResult};
+
+pub const DOI_DOMAIN: &str = "https://doi.org/";
+pub const ORCID_DOMAIN: &str = "https://orcid.org/";
+
+#[cfg_attr(
+    feature = "backend",
+    derive(DieselNewType, juniper::GraphQLScalarValue)
+)]
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct Doi(String);
+
+#[cfg_attr(
+    feature = "backend",
+    derive(DieselNewType, juniper::GraphQLScalarValue)
+)]
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct Orcid(String);
+
+#[cfg_attr(
+    feature = "backend",
+    derive(DieselNewType, juniper::GraphQLScalarValue)
+)]
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct Timestamp(DateTime<Utc>);
+
+impl Default for Doi {
+    fn default() -> Doi {
+        Doi {
+            0: Default::default(),
+        }
+    }
+}
+
+impl Default for Orcid {
+    fn default() -> Orcid {
+        Orcid {
+            0: Default::default(),
+        }
+    }
+}
+
+impl Default for Timestamp {
+    fn default() -> Timestamp {
+        Timestamp {
+            0: TimeZone::timestamp(&Utc, 0, 0),
+        }
+    }
+}
+
+impl fmt::Display for Doi {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", &self.0.replace(DOI_DOMAIN, ""))
+    }
+}
+
+impl fmt::Display for Orcid {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", &self.0.replace(ORCID_DOMAIN, ""))
+    }
+}
+
+impl fmt::Display for Timestamp {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", &self.0.format("%F %T"))
+    }
+}
+
+impl FromStr for Doi {
+    type Err = ThothError;
+
+    fn from_str(input: &str) -> ThothResult<Doi> {
+        use lazy_static::lazy_static;
+        use regex::Regex;
+        lazy_static! {
+        static ref RE: Regex = Regex::new(
+            // ^    = beginning of string
+            // (?:) = non-capturing group
+            // i    = case-insensitive flag
+            // $    = end of string
+            // Matches strings of format "[[http[s]://]doi.org/]10.XXX/XXX"
+            // and captures the identifier segment starting with the "10." directory indicator
+            // Corresponds to database constraints although regex syntax differs slightly
+            // (e.g. `;()/` do not need to be escaped here)
+            r#"^(?i:(?:https?://)?doi\.org/)?(10\.\d{4,9}/[-._;()/:a-zA-Z0-9]+$)"#).unwrap();
```

Comment:
Could we maybe allow more inputs here, e.g. https://www.doi.org/{} (mentioned in #252), https://dx.doi.org/{} (which was the norm until a few years ago)

Reply:
Yep, definitely - I should have thought of the `www` version, although I wasn't aware of the `dx` version. Are there any other variants I should account for?

Reply:
I _think_ those are the only other variants I've come across. We can potentially allow any subdomain just to be on the safe side...

Reply:
I'm inclined to err on the side of strictness, so I've just added the `dx` subdomain for now - if we hear about any others, we can tweak it later (and the workaround is easy/obvious enough).

Reply:
Fair enough